.github/scripts/filter_test_configs.py

```diff
@@ -0,0 +1,146 @@
+        return set()
+
+    return {label.get("name") for label in response.json() if label.get("name")}
+
+
+def filter(test_matrix: Dict[str, List[Any]], labels: Set[str]) -> Dict[str, List[Any]]:
+    """
+    Select the list of test config to run from the test matrix. The logic works
+    as follows:
+
+    If the PR has one or more labels as specified in the VALID_TEST_CONFIG_LABELS set, only
+    these test configs will be selected.  Note that regular ciflow labels like ciflow/trunk
+    don't count here, only test configs.
+
+    If the PR has none of the ciflow label, all tests are run as usual.
+    """
+
+    filtered_test_matrix: Dict[str, List[Any]] = {
+        "include": []
+    }
+
+    for entry in test_matrix.get("include", []):
+        config_name = entry.get("config", "")
+        if not config_name:
+            continue
+
+        label = f"{CIFLOW_PREFIX}{config_name.strip().lower()}"
+        if label in labels:
+            print(f"Select {config_name} because label {label} is presented in the pull request by the time the test starts")
+            filtered_test_matrix["include"].append(entry)
+
+    valid_test_config_labels = labels.intersection(VALID_TEST_CONFIG_LABELS)
+
+    if filtered_test_matrix["include"] or valid_test_config_labels:
+        # When the filter test matrix contain matches or if a valid test config label
+        # is found in the PR, return the filtered test matrix
+        return filtered_test_matrix
+    else:
+        # Found no valid label, return the same test matrix as before so that all
+        # tests can be run normally
+        return test_matrix
+
+
+def main() -> None:
+    args = parse_args()
+    # Load the original test matrix set by the workflow. Its format, however,
+    # doesn't follow the strict JSON format, so we load it using yaml here for
+    # its more relaxed syntax
+    test_matrix = yaml.safe_load(args.test_matrix)
+    pr_number = args.pr_number
+    tag = args.tag
+
+    # If the tag matches, we can get the PR number from it
+    tag_regex = re.compile(fr"^{CIFLOW_PREFIX}\w+/(?P<pr_number>\d+)$")
+
+    if not pr_number and not tag:
+        # This can be none or empty like when the workflow is dispatched manually
+        filtered_test_matrix = test_matrix
+
+    elif not pr_number:
+        m = tag_regex.match(tag)
+
+        if m:
+            pr_number = m.group("pr_number")
+        else:
+            # There is a tag but it isn't ciflow, so there is nothing left to do
+            filtered_test_matrix = test_matrix
+
+    if pr_number:
+        # First, query all the labels from the pull requests
+        labels = get_labels(int(pr_number))
+        # Then filter the test matrix and keep only the selected ones
+        filtered_test_matrix = filter(test_matrix, labels)
+
+    # Set the filtered test matrix as the output
+    print(f"::set-output name=test-matrix::{json.dumps(filtered_test_matrix)}")
```

Comment:
What does this `::set-output` do?

Reply:
Tells GHA to treat it as output :)
https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#example-setting-a-value

Reply:
Exactly. GHA parses this and assign the string value to the output variable name accordingly.  Later on, you can see these variables are referred to in the workflow, i.e.

```
# The syntax is steps.STEP_ID.outputs.VARIABLE_NAME
test-matrix: ${{ steps.filter.outputs.test-matrix }}
is-test-matrix-empty: ${{ steps.filter.outputs.is-test-matrix-empty }}
```